tools/swift-stdlib-tool/swift-stdlib-tool.cpp

```diff
@@ -670,583 +642,568 @@ -(NSString *)sst_command {
+            unsigned_dst_dir = unsigned_dst_dir/platform;
         }
-        
-        // If the user specifies --strip-bitcode but not --sign, this
-        // will cause the dylibs to get copied, stripped, but not resigned.
-        // This will cause apps to fail to launch because the code signature
-        // is invalid.  In this case, ignore --strip-bitcode.
-        if (stripBitcode && !ident) {
-            stripBitcode = false;
+    }
+    
+    // If the user specifies --strip-bitcode but not --sign, this
+    // will cause the dylibs to get copied, stripped, but not resigned.
+    // This will cause apps to fail to launch because the code signature
+    // is invalid.  In this case, ignore --strip-bitcode.
+    if (stripBitcode && ident.empty()) {
+        stripBitcode = false;
+    }
+
+    // Collect executables from the --scan-folder locations.
+    for (auto const &embedDir : embedDirs) {
+        for (auto const &entry : std::filesystem::recursive_directory_iterator(embedDir)) {
+            if (entry.exists() && !entry.is_directory() &&
+                0 == access(entry.path().c_str(), X_OK))
+            {
+                executables.push_back(entry.path());
+            } else {
+                log_vv("%s is not an executable file", entry.path().c_str());
+            }
         }
+    }
+         
+    // Collect Swift library names from the input files.
+    // If the library does not exist in src_dir then assume the user wrote 
+    // their own library named libswift* and is handling it elsewhere.
+    __block std::unordered_map<std::string,
+                    std::unordered_set<std::string>> swiftLibs;
+    for (auto const &path : executables) {
+        process(path,
+           ^(std::filesystem::path const &linkedLib) {
+                auto const linkedSrc = src_dir/linkedLib;
+                if (std::filesystem::exists(linkedSrc)) {
+                    swiftLibs[linkedLib] = std::unordered_set<std::string>();
+                }
+            }, 
+            NULL);
+    }
 
-        // Collect executables from the --scan-folder locations.
-        for (NSString *embedDir in embedDirs) @autoreleasepool {
-            NSDirectoryEnumerator *dir = [fm enumeratorAtPath:embedDir];
-            for (NSString *p in dir) @autoreleasepool {
-                BOOL isDir;
-                NSString *p2 = [embedDir stringByAppendingPathComponent:p];
-                if ([fm fileExistsAtPath:p2 isDirectory:&isDir]  &&  
-                    !isDir  &&  
-                    [fm isExecutableFileAtPath:p2])
+    // Collect more Swift library names from the Swift libraries themselves.
+    // Also collect the Swift libraries' UUIDs.
+    __block std::vector<std::string> worklist;
+    worklist.reserve(swiftLibs.size());
+    for (auto const &[lib, _] : swiftLibs) {
+        worklist.push_back(lib);
+    }
+    while (worklist.size()) {
+        auto const &lib = worklist.back();
+        worklist.pop_back();
```

Comment:
You cannot grab a reference to the `back()` and immediately `pop_back()`. It will invalidate the reference. It probably works today because of luck, but it is not guaranteed to work. It is probably safer to remove the reference and pay the price of the copy.